aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -933,6 +934,53 @@ Tensor _test_serialization_subcmul(const Tensor& self, const Tensor& other, Scal
   return self - (other * alpha);
 }
 
+Tensor& heaviside_tensor_out(Tensor& result, const Tensor& self, const Tensor& val) {
+  TORCH_CHECK(!self.is_complex() && !result.is_complex() && !val.is_complex(),
+              "heaviside is not implemented for complex tensors.");
+
+  auto iter = TensorIterator::binary_op(result, self, val.to(self.dtype()), /*check_mem_overlap=*/true);
```

Comment:
This seems odd.

If self is an int64 tensor, for example, and val is a float tensor, then this will truncate all the float values. I don't know if a user would find that natural.

This could require that self and val have the same dtype for now to avoid confusion? Another option would be to promote based on the dtypes of self and val like `np.clip` does. Then an int64 self and a float val would produce a float result, for example.

Reply:
I agree, and for the time being, I'd like to keep the same dtype for self and val to avoid confusion.